### PR TITLE
Snowflake upperrel pushdown

### PR DESCRIPTION
**Pull request by gruuya**
Implement aggregation pushdown for SQLAlchemy supported FDWs, principally with the aim of supporting Snowflake.

CU-23kfzuz

### CODE REVIEW

File: python/multicorn/sqlalchemyfdw.py [hunk 1]
```diff
@@ -440,20 +478,21 @@ def _build_statement(self, quals, columns, sortkeys):
...
+        statement = self._build_statement(quals, columns, sortkeys, aggs=aggs, group_clauses=group_clauses)
         log_to_postgres(str(statement), DEBUG)
 
         # If a dialect doesn't support streaming using server-side cursors,
         # we hack it a bit by sending LIMIT/OFFSET queries.
         offset = 0
         with inject_envvars(self.envvars):
             while True:
-                if self.batch_size is not None:
+                if self.batch_size is not None and not is_aggregation:
```
Comment (mildbyte) [inline python/multicorn/sqlalchemyfdw.py:495]: Do we really need to disable pagination when aggregating (i.e. does `SELECT agg(x), y FROM z GROUP BY y LIMIT 10 OFFSET 20` not work properly with aggregations)? If you have a test with a small `batch_size` (smaller than the aggregation's return size), does it break with this change removed?

File: python/multicorn/sqlalchemyfdw.py [hunk 2]
```diff
@@ -401,32 +410,61 @@ def can_sort(self, sortkeys):
...
+            "agg_functions": list(_PG_AGG_FUNC_MAPPING),
+            "operators_supported": [op for op in OPERATORS if isinstance(op, str)],
+        }
+
+    def explain(self, quals, columns, sortkeys=None, aggs=None, group_clauses=None, verbose=False):
         sortkeys = sortkeys or []
-        statement = self._build_statement(quals, columns, sortkeys)
-        return [str(statement)]
+        statement = self._build_statement(quals, columns, sortkeys, aggs=aggs, group_clauses=group_clauses)
+        return ["\n" + str(statement.compile(dialect=self.engine.dialect, compile_kwargs={"literal_binds": True})) + "\n"]
```
Comment (mildbyte) [inline python/multicorn/sqlalchemyfdw.py:423]: This looks kinda scary (https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#rendering-bound-parameters-inline) -- a comment here mentioning why it's required would be nice.

**Review by mildbyte — COMMENTED** [review]
This LGTM, assuming the tests in the main repo pass!

File: python/multicorn/sqlalchemyfdw.py (hunk 1 above)
Comment (gruuya) [reply python/multicorn/sqlalchemyfdw.py:495]: No need really, I just left it like that to simplify debugging edge cases and forgot about it.

Fwiw, I did test it with batching turned on, and a small batch size option, and it was correct, bringing it back now.

File: python/multicorn/sqlalchemyfdw.py (hunk 2 above)
Comment (gruuya) [reply python/multicorn/sqlalchemyfdw.py:423]: The sole reason for this was to assert actual values used in the remote query, though given that the query results are also being asserted it may be an overkill.

Comment (gruuya) [reply python/multicorn/sqlalchemyfdw.py:427]: I've left a comment expanding on the reason for this.

**Review by mildbyte — APPROVED** [review]
